### DISCUSSION TITLE
gRPC in Next.js

### DISCUSSION THREAD

**Discussion by SabariVig** [vercel/next.js#13862] [Help] [answered]
i tried grpc-web i got an error stating `ReferenceError: XMLHttpRequest ` is not defined  it did not work,
 I tried grpc-node it also did not work 
is there a way to get grpc working in next js

**Comment by Multiply** [comment]
We ended up using improbable's grpc-web implementation because it works both on the server and in the browser at the same time.

**Reply by SabariVig** [reply]
But when tried I get `ReferenceError: XMLHttpRequest ` as mentioned above. Is there any examples where i can refer or any blog post doing the same

**Reply by amantri** [reply]
Next.js uses server side rendering. Wrapping the grpc client call in a useEffect hook fixed the XMLHttpRequest for me.

**Comment by Multiply** [comment, marked as answer]
On the server we do the following before using grpc-web:
```
import { grpc } from '@improbable-eng/grpc-web'
import { NodeHttpTransport } from '@improbable-eng/grpc-web-node-http-transport'

grpc.setDefaultTransport(NodeHttpTransport())
```

**Reply by Kupstas** [reply]
You're taking about https://github.com/improbable-eng/grpc-web. But it's not working for https://github.com/grpc/grpc-web

**Comment by j-mcgregor** [comment]
Does anyone have an example of a NextJS app with gRPC implementation? I'm finding the docs confusing

**Reply by SabariVig** [reply]
I managed to get it working 
Check out https://github.com/SabariVig/fullstack-grpc if you need any reference.

**Reply by lucksp** [reply]
To run with Vercel, do we need to setup docker & custom server?

**Comment by willemmulder** [comment]
An old thread, but I just came across https://github.com/Janpot/next-rpc which looks really awesome! Did anyone try that?
